examples/add/train_add_distill_sdxl_wds.py

```diff
@@ -0,0 +1,1773 @@
+# TODO: implement image conditioning (see Section 3.2 of paper)
+class Discriminator(torch.nn.Module):
+    """
+    StyleGAN-T-style discriminator for adversarial diffusion distillation (ADD).
+    """
+
+    def __init__(
+        self,
+        pretrained_feature_network: str = "vit_small_patch16_224_dino",
+        cond_embedding_dim: int = 512,
+        patch_size: List[int] = [16, 16],
+        hooks: List[int] = [2, 5, 8, 11],
+        start_index: int = 1,
+    ):
+        super().__init__()
+        self.cond_embedding_dim = cond_embedding_dim
+
+        # Frozen feature network, e.g. DINO
+        self.feature_network = FeatureNetwork(
+            pretrained_feature_network=pretrained_feature_network,
+            patch_size=patch_size,
+            hooks=hooks,
+            start_index=start_index,
+        )
+
+        # Trainable discriminator heads
+        heads = []
+        for i in range(self.feature_network.num_hooks):
+            heads += [str(i), DiscriminatorHead(self.feature_network.embed_dim, cond_embedding_dim)]
+        self.heads = torch.nn.ModuleDict(heads)
+
+    def train(self, mode: bool = True):
+        self.feature_network = self.feature_network.train(False)
+        self.heads = self.heads.train(mode)
+        return self
+
+    def eval(self):
+        return self.train(False)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        c: torch.Tensor,
+        transform_positive=True,
+        return_dict: bool = True,
+    ):
+        # TODO: do we need the augmentations from the original StyleGAN-T code?
+        if transform_positive:
+            # Transform to [0, 1].
+            x = x.add(1).div(2)
+
+        # Forward pass through feature network.
+        features = self.feature_network(x)
+
+        # Apply discriminator heads.
+        logits = []
+        for k, head in self.heads.items():
+            logits.append(head(features[k], c).view(x.size(0), -1))
+        logits = torch.cat(logits, dim=1)
+
+        if not return_dict:
+            return (logits,)
+
+        return DiscriminatorOutput(logits=logits)
+
+
+def log_validation(vae, unet, args, accelerator, weight_dtype, step, name="student"):
+    logger.info("Running validation... ")
+
+    unet = accelerator.unwrap_model(unet)
+    pipeline = StableDiffusionXLPipeline.from_pretrained(
+        args.pretrained_teacher_model,
+        vae=vae,
+        unet=unet,
+        revision=args.revision,
+        torch_dtype=weight_dtype,
+    )
+    pipeline = pipeline.to(accelerator.device)
+    pipeline.set_progress_bar_config(disable=True)
+
+    if args.enable_xformers_memory_efficient_attention:
+        pipeline.enable_xformers_memory_efficient_attention()
+
+    if args.seed is None:
+        generator = None
+    else:
+        generator = torch.Generator(device=accelerator.device).manual_seed(args.seed)
+
+    validation_prompts = [
```

Comment:
you can use a prompt dataset here

Reply:
For a reference example script, that ain't necessary.

Reply:
my suggestion was based on the other training example scripts :-)